### ISSUE TITLE
Where should be caching?

### ISSUE THREAD

**Issue by IssamT** [JuliaOpt/MathOptInterface.jl#106]
I agree that the current strategy of JuMP is useful but I don't like the idea of cache being only in JuMP... I understand and accept the argument of having MOI a pure API but It would be nice to have the MOIU instance shared by JuMP and the solver wrapper. Example: if a solver wrapper wants to answer the query about MOI.ConstraintSet it will need to store it's own copy or build the set on the fly which would be a waste since the set exists and is stored in JuMP.

The problem with sharing is that I don't know if it is possible with julia to deny the solvers from modifying the instance once it is shared.

One possible solution: given a valid MOIU instance there are queries that do not need a solver to be answered. so wrappers do not implement those methods and we move them to MOIU. Everything using MOIU instance goes to MOIU to keep it separated from what solvers should implement. Then I see methods in MOIU like

````julia
MOIU.addconstraint!(m::MOI.AbstractSolverInstance, func::F, set::S, <:MOIU.AbstractInstance{T}, updateform::Bool)
#MOIU update its instance and if updateform == false
#MOIU keeps track of what is not yet passed through vectors for each combination F,S as discussed
#here https://github.com/JuliaOpt/MathOptInterface.jl/issues/63#issuecomment-317526848 

MOIU.update_solver_instance() 
#calls MOI.addconstraints! and deletes the vectors but keeps the instance.

MOIU.getattribute(m::MOIU.AbstractInstance{T}, ::MOI.ConstraintSet}, ref)
#replacing what was before in MOI and implemented by the solvers 
#MOI.getattribute(m::MOI.SolverInstance{T}, ::MOI.ConstraintSet}, ref)
````

As @blegat mentioned, this will not work in the following case:
if the solver wrapper wants to have a mode on the fly where every info required for a single constraint is added and deleted right afterwards (being only kept inside the solver). 

Such solver wrappers will be denied to have an answer to the following query since now it s the responsability of MOIU instance. 

````julia
MOI.getattribute(m::MOI.SolverInstance{T}, ::MOI.ConstraintSet}, ref)
````

To solve this issue, solvers need to have access to the instance (bad thing). and all those methods will not be replaced but only get a default implementation that uses MOIU instance (when available):

So Instead of this

````julia
MOIU.getattribute(m::MOIU.AbstractInstance{T}, ::MOI.ConstraintSet, ref)
# replacing what was before in MOI and implemented by the solvers
# MOI.getattribute(m::MOI.SolverInstance{T}, ::MOI.ConstraintSet, ref)
````

We can have this

````julia
MOIU.getattribute(m::MOIU.AbstractInstance{T}, ::MOI.ConstraintSet, ref)
# helps providing default implementation to
# MOI.getattribute(m::MOI.SolverInstance{T}, ::MOI.ConstraintSet}, ref)
function MOI.getattribute(m::MOI.AbstractSolverInstance{T}, s::MOI.ConstraintSet, ref)
    if has_instance(m)
       MOIU.getattribute(instance(m), s, ref)
    else
      throw MethodError()
    end
end
````

With JuMP being now only one among other possible users of MOI / MOIU doing instead of this :

````julia
functions1 = Vector{FunctionSetA}()
sets1 = Vector{ConstraintSetA}()
# fill the vectors
functions2 = Vector{FunctionSetA}()
sets2 = Vector{ConstraintSetB}()
# fill the vectors
# ...continue for each F,S type combination
MOI.addconstraints!(m,functions1, sets1)
MOI.addconstraints!(m,functions2, sets2)
#...continue for each F,S type combination
````

the follwing:

````julia
MOIU.addconstraint!(m, f, s, false) # assuming instance(m) returns an MOIU instance
#...continue adding the other constraints
MOIU.update_solver_instance(m)
````

**Comment by IssamT** [comment]
related issue : https://github.com/JuliaOpt/MathOptInterface.jl/issues/63#issuecomment-317526848

**Comment by mlubin** [comment]
@IssamT is your main concern just with ``ConstraintSet`` and ``ConstraintFuncion``? There's really no need for the MOI wrapper to implement these if it's a difficult operation. They're not very common requests, but *some* solver APIs support them, so they're part of the interface.

What's the use case where it would be convenient for the solver to cache? If using JuMP, JuMP can just request this information from its instance instead of the solver.

**Comment by mlubin** [comment]
What it sounds like you want is to remove the "attaching an instance to a solver" logic from JuMP and move it into somewhere else. That's could potentially be a good idea.

**Comment by IssamT** [comment]
There are other attributes: ObjectiveFunction, ObjectiveSense, NumberOfConstraints (for all combinations), ConstraintList,  ... 

However, attributes are not my main concern. My main concern is the caching of constraint to push them in batches to the solver

**Comment by mlubin** [comment]
Why is it a problem if ``addconstraint!`` and ``addconstraints!`` have the same performance characteristics as they would in the solver's own API?

**Comment by IssamT** [comment]
Yes I would be happy to remove the "attaching an instance to a solver" logic from JuMP. It's mostly for modularity. As @blegat said, some solvers might internally call other solvers without needing JuMP

**Comment by IssamT** [comment]
In some cases (MILP solvers for example) we often want to use addconstraints! whether using MOI or the Solver own API. So I think if there is a module used by JuMP, better have it available to other users making it easy for them to push the constraint in batches

**Comment by IssamT** [comment]
> Why is it a problem if addconstraint! and addconstraints! have the same performance characteristics as they would in the solver's own API?

There is no problem with that on the level on MOI. I like it because it s a pure API and that s what the wrappers should implement. Notice that I used **MOIU**.addconstraint! and **MOIU**.update_solver_instance! which are utility functions.

**Comment by mlubin** [comment]
We can start a discussion on what the instance-attached-to-solver object would look like.

**Comment by blegat** [comment]
Below is a suggestion for the interface, let me know what you think
```julia
# Creates an Instance and further calls to addconstraint!, ... will be handled by the Instance and won't go through to the solver
function MOI.enablecache(::AbstractSolverInstance) end # Does nothing by default
# Apply the changes to the solver and further calls to addconstraint!, ... will be handled by the Instance and won't go through to the solver
function MOI.applycache(::AbstractSolverInstance) end # Does nothing by default
# Drop the Instance and further calls to addconstraint!, ... will be handled by the solver
function MOI.disablecache(::AbstractSolverInstance) end # Does nothing by default
```

**Comment by mlubin** [comment]
I'll put some thoughts down before I go away. Probably won't have a conclusion on this before I return.

To summarize my view of the discussion so far, from the perspective of JuMP, I just want MOI wrappers to be direct wrappers of the solver API, and if there's a modification that's not supported then I can just toss the solver instance and create a new once. I don't care about caching on the solver side because JuMP will always have a copy of the instance saved. A JuMP model can exist attached or unattached to a solver instance. If we want to take this concept of a JuMP model and move this basic functionality into an MOI-level object that's okay with me since it makes JuMP simpler and more modular. I think everyone is in agreement up to here.

A good sanity check for the design of this new object, let's call it ``Instance++`` is to ask the question of if it could support being attached to more than one solver instance at the same time. The instance data should be the central object and the solver secondary. We might want to allow a JuMP model to be attached to multiple solvers at some point in the future.

Now there's a second question which seems quite different, which is what if the MOI wrapper wants to use a stored `Instance` to improve its performance. @IssamT mentioned that it would be useful for non-JuMP users for the MOI wrapper to cache constraints before sending them to the solver. I don't think ``Instance++`` solves this problem at all because ``Instance++`` has to be generic with respect to how it caches data, while each solver API might have very different performance characteristics. One caching strategy might be reasonable for Gurobi but unreasonable for GLPK. I think MOI wrappers should be allowed to keep caches like this, but there should be an option to disable it (possibly disabled by default) because it can have significant memory usage implications for JuMP users.

**Comment by IssamT** [comment]
By caching I only meant that we pass all the modification of the problem at once in a simple generic, solver-independent way. It can only improve the performance since in the worst case a wrapper can apply them one by one getting the same performance as without caching. Maaaaybe if the caching wasn't generic but rather done in a way that favors some solver, it can be better for that solver but this type of caching will require a lot more work.

@mlubin, I think that even the simple generic caching adds a bit of complexity. And it is quite orthogonal to the other MOI issues. So we can drop it for now. I will reopen another issue once I have some clear performance differences.

closed in favor of #32
